api/clients/v2/payload_disperser.go

```diff
@@ -0,0 +1,222 @@
+	defer cancel()
+	blobStatusReply, err := pd.pollBlobStatus(timeoutCtx, blobKey, blobStatus.ToProfobuf())
+	if err != nil {
+		return nil, fmt.Errorf("poll blob status: %w", err)
+	}
+	pd.logger.Debug("Blob status CERTIFIED", "blobKey", blobKey)
+
+	timeoutCtx, cancel = context.WithTimeout(ctx, pd.config.ContractCallTimeout)
+	defer cancel()
+	nonSignerStakesAndSignature, err := pd.certVerifier.GetNonSignerStakesAndSignature(
+		timeoutCtx, blobStatusReply.GetSignedBatch())
+	if err != nil {
+		return nil, fmt.Errorf("get non signer stake and signature: %w", err)
+	}
+	pd.logger.Debug("Retrieved NonSignerStakesAndSignature", "blobKey", blobKey)
+
+	eigenDACert, err := verification.BuildEigenDACert(
+		blobStatusReply.GetBlobInclusionInfo(),
+		blobStatusReply.GetSignedBatch().GetHeader(),
+		nonSignerStakesAndSignature)
+	if err != nil {
+		return nil, fmt.Errorf("build eigen da cert: %w", err)
+	}
+	pd.logger.Debug("Constructed EigenDACert", "blobKey", blobKey)
+
+	timeoutCtx, cancel = context.WithTimeout(ctx, pd.config.ContractCallTimeout)
+	defer cancel()
+	err = pd.certVerifier.VerifyCertV2(timeoutCtx, eigenDACert)
+	if err != nil {
+		return nil, fmt.Errorf("verify cert for blobKey %v: %w", blobKey, err)
+	}
+	pd.logger.Debug("EigenDACert verified", "blobKey", blobKey)
+
+	return eigenDACert, nil
+}
+
+// Close is responsible for calling close on all internal clients. This method will do its best to close all internal
+// clients, even if some closes fail.
+//
+// Any and all errors returned from closing internal clients will be joined and returned.
+//
+// This method should only be called once.
+func (pd *PayloadDisperser) Close() error {
+	err := pd.disperserClient.Close()
+	if err != nil {
+		return fmt.Errorf("close disperser client: %w", err)
+	}
+
+	return nil
+}
+
+// pollBlobStatus polls the disperser for the status of a blob that has been dispersed
+//
+// This method will only return a non-nil BlobStatusReply if the blob is reported to be CERTIFIED prior to the timeout.
+// In all other cases, this method will return a nil BlobStatusReply, along with an error describing the failure.
+func (pd *PayloadDisperser) pollBlobStatus(
+	ctx context.Context,
+	blobKey core.BlobKey,
+	initialStatus dispgrpc.BlobStatus,
+) (*dispgrpc.BlobStatusReply, error) {
+
+	previousStatus := initialStatus
+
+	ticker := time.NewTicker(pd.config.BlobStatusPollInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil, fmt.Errorf(
+				"timed out waiting for %v blob status, final status was %v: %w",
+				dispgrpc.BlobStatus_CERTIFIED,
+				previousStatus,
+				ctx.Err())
+		case <-ticker.C:
+			// This call to the disperser doesn't have a dedicated timeout configured.
+			// If this call fails to return in a timely fashion, the timeout configured for the poll loop will trigger
+			blobStatusReply, err := pd.disperserClient.GetBlobStatus(ctx, blobKey)
```

Comment:
don't think this is true. call is blocking afaiu, so no other branches of the select loop can run in parallel. but the context will fail at some point once the context is done which will amount to the same thing

Reply:
also it does have a timeout, its the one configured and passed in to this function as part of ctx

Reply:
> also it does have a timeout, its the one configured and passed in to this function as part of ctx

This is what the comment is attempting to convey. The spot in the code intentionally doesn't enforce an *additional* timeout for the `GetBlobStatus` call, because the timeout configured and passed in as a parameter functions to break out of the polling loop, AND abort this call to the disperser if necessary.

Does that make sense? If so, do you have any suggestions for how to word this in a way that would be more clear?

Reply:
I see. That's fine with me for now. sorry for bike-shedding.